File: Data-Structures-and-Algorithms/0202-HappyNumber/isHappy.py
class Solution(object):
    def isHappy(self, n):
        seen_before = set()

        while n != 1:
            n = sum([int(d)**2 for d in str(n)])
            if n in seen_before:
                return False
            else:
                seen_before.add(n)
        
        return True

n = 19
print(Solution().isHappy(n))

# O(logn) with log base of 10 (number of digits scales in log10 with the number)
# O(logn) for space
# Sum of each digit squared of a number, recursive, until number is 1 

File: Data-Structures-and-Algorithms/0202-HappyNumber/isHappy.js
const isHappy = function(n) {

    const nextnum = function(n) {
        let sum = 0;
        var arr = Array.from(String(n), Number);
        for (var i=0; i<arr.length; i++) {
            sum += (arr[i]*arr[i]);
        }
        return sum
    }

    slow = n;
    fast = nextnum(n);
    while (fast != 1 && slow != fast) {
        slow = nextnum(slow);
        fast = nextnum(nextnum(fast));
    }

    return fast == 1
};

n = 19
console.log(isHappy(n))

// O(logN) with log base 10
// O(1) since no extra hashset or hashmap needed
// Sum of each digit squared of a number, recursive, until number is 1 